В данном варианте используется детектор особых точек SIFT. Для построения карты глубины используется нейросетевая модель MiDAs.
Для восстановления позы камеры используется функция cv2.solvePnPRansac. 
Из видео выбран каждый 10 кадр, для существенного смещения камеры.

In [2]:
import cv2
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
from transformers import DPTForDepthEstimation, DPTFeatureExtractor
import torch
import os
from tqdm import tqdm



d:\soft\Anaconda\envs\DZ3\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [3]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model = DPTForDepthEstimation.from_pretrained("Intel/dpt-large").to(device)
feature_extractor = DPTFeatureExtractor.from_pretrained("Intel/dpt-large")


Some weights of DPTForDepthEstimation were not initialized from the model checkpoint at Intel/dpt-large and are newly initialized: ['neck.fusion_stage.layers.0.residual_layer1.convolution1.bias', 'neck.fusion_stage.layers.0.residual_layer1.convolution1.weight', 'neck.fusion_stage.layers.0.residual_layer1.convolution2.bias', 'neck.fusion_stage.layers.0.residual_layer1.convolution2.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
d:\soft\Anaconda\envs\DZ3\Lib\site-packages\transformers\models\dpt\feature_extraction_dpt.py:28: FutureWarning: The class DPTFeatureExtractor is deprecated and will be removed in version 5 of Transformers. Please use DPTImageProcessor instead.
  warnings.warn(


In [4]:
# --- Функция для получения карты глубины ---
def estimate_depth(image):
    inputs = feature_extractor(images=image, return_tensors="pt").to(device)
    with torch.no_grad():
        outputs = model(**inputs)
        depth = outputs.predicted_depth[0].cpu().numpy()
    depth_rescaled = (depth - depth.min()) / (depth.max() - depth.min())  # Нормализация
    return depth_rescaled

In [6]:
# --- Функция для восстановления движения камеры ---
def compute_camera_motion(kp1, kp2, matches, depth_map1, K):
    points3D_1 = []
    points2D_2 = []
    
    h, w = depth_map1.shape  # Размер карты глубины
    
    for match in matches:
        u1, v1 = int(kp1[match.queryIdx].pt[0]), int(kp1[match.queryIdx].pt[1])
        u2, v2 = int(kp2[match.trainIdx].pt[0]), int(kp2[match.trainIdx].pt[1])
        
        # Проверка, что координаты находятся в пределах карты глубины
        if 0 <= u1 < w and 0 <= v1 < h:
            Z1 = depth_map1[v1, u1]
            if Z1 > 0:  # Проверка валидности глубины
                X1 = (u1 - K[0, 2]) * Z1 / K[0, 0]
                Y1 = (v1 - K[1, 2]) * Z1 / K[1, 1]
                points3D_1.append([X1, Y1, Z1])
                points2D_2.append([u2, v2])
    
    if len(points3D_1) == 0 or len(points2D_2) == 0:
        raise ValueError("Недостаточно валидных точек для восстановления движения камеры.")
    
    points3D_1 = np.array(points3D_1)
    points2D_2 = np.array(points2D_2, dtype=np.float32)

    _, rvec, tvec, _ = cv2.solvePnPRansac(points3D_1, points2D_2, K, None)
    R, _ = cv2.Rodrigues(rvec)
    return R, tvec

In [7]:
from mpl_toolkits.mplot3d import Axes3D

def create_trajectory(poses):
    trajectory = [np.array([0, 0, 0])]
    current_pose = np.eye(4)
    pose_cam = [np.array([[0, 0, 0],[0, 0, 0],[0, 0, 0]])]

    for R, t in poses:
        T = np.eye(4)
        T[:3, :3] = R
        T[:3, 3] = t.T
        current_pose = np.dot(current_pose, T)
        trajectory.append(current_pose[:3, 3])
        pose_cam.append(current_pose[:3, :3])

    return np.array(trajectory),np.array(pose_cam)

In [8]:
import plotly.graph_objects as go
import numpy as np
# Создание фигуры

def plot_trajectory_3D(trajectory, pose_cam):
    fig = go.Figure()


    fig.add_trace(go.Scatter3d(
        x=trajectory[:, 0],
        y=trajectory[:, 1],
        z=trajectory[:, 2],
        mode='lines+markers',
        marker=dict(size=5, color='blue'),
        line=dict(color='blue')
    ))

    # Добавление ориентации камеры в каждой точке траектории
    for i, (R, t) in enumerate(zip(pose_cam,trajectory)):
        # Направление камеры (ось Z камеры)
        camera_direction = R @ np.array([0, 0, 1])  # Направление оси Z камеры
        camera_direction_end = t + camera_direction * 0.5  # Конец вектора направления

        # Добавляем линию, представляющую направление камеры
        fig.add_trace(go.Scatter3d(
            x=[t[0], camera_direction_end[0]],
            y=[t[1], camera_direction_end[1]],
            z=[t[2], camera_direction_end[2]],
            mode='lines',
            line=dict(color='green', width=2),
            name=f'Camera Direction {i}' if i == 0 else None,
            showlegend=False if i > 0 else True
        ))

    # Настройка макета графика
    fig.update_layout(
        title='Camera Motion Trajectory',
        scene=dict(
            xaxis_title='X-axis',
            yaxis_title='Y-axis',
            zaxis_title='Z-axis'
        ),
        showlegend=True
    )

    # Показать график
    fig.show()

In [9]:
# Получить список файлов с изображениями
image_folder = "./frames"
image_files = sorted([os.path.join(image_folder, f) for f in os.listdir(image_folder) if f.endswith(".png")])

In [11]:
K = np.array([[3000 ,   0.    , 960],
       [  0.    , 3000 , 540 ],
      [  0.    ,   0.    ,   1.]], dtype=np.float32) # Матрица камеры
sift = cv2.SIFT_create()
bf = cv2.BFMatcher()
trajectory = [[0, 0, 0]]
T_global = np.eye(4)
poses = []

for i in tqdm(range(len(image_files) - 1)):
   try:
         img1 = cv2.imread(image_files[i])
         img2 = cv2.imread(image_files[i + 1])
        # img1 = cv2.resize(img1, (img1.shape[1] // 2, img1.shape[0] // 2), interpolation=cv2.INTER_AREA)
        # img2 = cv2.resize(img2, (img2.shape[1] // 2, img2.shape[0] // 2), interpolation=cv2.INTER_AREA)  
        

         depth_map1 = estimate_depth(img1)
         kp1, des1 = sift.detectAndCompute(cv2.cvtColor(img1, cv2.COLOR_BGR2GRAY), None)

       
 
         # Обработка текущего кадра
         depth_map2 = estimate_depth(img2)
         kp2, des2 = sift.detectAndCompute(cv2.cvtColor(img2, cv2.COLOR_BGR2GRAY), None)

         # Сопоставление ключевых точек
         matches = bf.knnMatch(des1, des2, k=2)
         good_matches = [m for m, n in matches if m.distance < 0.35 * n.distance]

         # Восстановление движения камеры
         R, t = compute_camera_motion(kp1, kp2, good_matches, depth_map1, K)
         poses.append((R, t))
         # print (poses)

         # Переход к следующему кадру
         depth_map1 = depth_map2
         kp1, des1 = kp2, des2
         #print(frame_counter)


         #trajectory = np.array(trajectory)
         
         
   except:
      pass
   
trajectory,pose_cam = create_trajectory(poses)

plot_trajectory_3D(trajectory,pose_cam)
   

100%|██████████| 113/113 [06:35<00:00,  3.50s/it]


Способ показал более точный результат, время обработки 113 кадров - 07:33.